### Data Merge

In [1]:
### Author: Anuvrat Chaturvedi
### Date: 13-Feb-2024
### Purpose: Merges all the input files to create combined dataframes (primary i.e. ex-Florida and Florida).

##### Declarations

In [2]:
#Importing utilities file with library imports and helper functions
%run "utils.ipynb"

In [3]:
#Loading input pickle files
df_income_1901_edited=pd.read_pickle(input_path+"\\df_income_1901_edited.pkl")
df_health_outcomes_edited=pd.read_pickle(input_path+"\\df_health_outcomes_edited.pkl")
df_sdoh_edited=pd.read_pickle(input_path+"\\df_sdoh_edited.pkl")
df_state_politics_edited=pd.read_pickle(input_path+"\\df_state_politics_edited.pkl")
df_state_public_spend_edited=pd.read_pickle(input_path+"\\df_state_public_spend_edited.pkl")
df_acs_edited=pd.read_pickle(input_path+"\\df_acs_edited.pkl")
df_zip2st_edited=pd.read_pickle(input_path+"\\df_zip2st_edited.pkl")
df_urban_rural_edited=pd.read_pickle(input_path+"\\df_urban_rural_edited.pkl")

In [4]:
#Printing shapes to check which file needs most attention for data compression
for df in [df_income_1901_edited, df_health_outcomes_edited, df_sdoh_edited, df_state_politics_edited, 
           df_state_public_spend_edited, df_acs_edited, df_urban_rural_edited]:
    print(df.shape)

(33774, 15)
(1169992, 10)
(291024, 4)
(57, 3)
(255, 3)
(33120, 14)
(33178, 4)


##### ZIP level data

In [5]:
[df_income_1901_edited.ZIP.nunique(), df_health_outcomes_edited.ZIP.nunique(), df_sdoh_edited.ZIP.nunique(), 
 df_acs_edited.ZIP.nunique(), df_urban_rural_edited.ZIP.nunique()]

[33774, 32409, 32336, 33120, 33178]

In [6]:
df_income_1901_essential=df_income_1901_edited[['ZIP', 'Estimate Households Total', 'Percent Households lt 10k', 
                                                'Estimate Households Median income (dollars)', 'Estimate Families Total',
                                                'Estimate Families Median income (dollars)', 'Estimate Nonfamily households Total',
                                                'Estimate Nonfamily households Median income (dollars)', 'Percent Households that are Families']]
df_income_1901_essential.shape

(33774, 9)

In [7]:
# Selecting essential columns from df_health_outcomes_edited and reshaping it to wide format with pivot
df_health_outcomes_essential = df_health_outcomes_edited[
    ["ZIP", "TotalPopulation", "Short_Question_Text", "Data_Value"]
]

df_health_outcomes_essential_wide = (
    df_health_outcomes_essential.pivot(
        index=["ZIP", "TotalPopulation"],
        columns="Short_Question_Text",
        values="Data_Value",
    )
    .reset_index()
    .rename_axis(None, axis=1)
)


# Uncomment to display the first few rows of the reshaped DataFrame
# display(df_health_outcomes_essential_wide.head())
# Displaying the shape of the reshaped DataFrame

df_health_outcomes_essential_wide.shape

(32409, 39)

In [8]:
# Reshaping df_sdoh_edited to wide format with pivot, using 'ZIP' and 'TotalPopulation' as index, 'Measure' as columns, and 'Data_Value' as values
df_sdoh_essential_wide = (
    df_sdoh_edited.pivot(
        index=["ZIP", "TotalPopulation"], columns="Measure", values="Data_Value"
    )
    .reset_index()
    .rename_axis(None, axis=1)
)


# Uncomment to display the first few rows of the reshaped DataFrame
# display(df_sdoh_essential_wide.head())

# Displaying the shape of the reshaped DataFrame
df_sdoh_essential_wide.shape

(32336, 11)

##### State level data

In [9]:
# Checking if any of the state level data needs to be pivoted
print(
    "df_state_politics_edited",
    df_state_politics_edited.shape,
    df_state_politics_edited.State.nunique(),
)
print("df_zip2st_edited", df_zip2st_edited.shape, df_zip2st_edited.ZIP.nunique())
print(
    "df_state_public_spend_edited",
    df_state_public_spend_edited.shape,
    df_state_public_spend_edited["State Name"].nunique(),
)

df_state_politics_edited (57, 3) 57
df_zip2st_edited (39368, 3) 39368
df_state_public_spend_edited (255, 3) 51


In [10]:
# Turns out state_public_spend_edited will need to be pivoted as it has multiple values of year for each state
df_state_public_spend_essential_wide = (
    df_state_public_spend_edited.pivot(
        index=["State Name"], columns="TimeFrame", values="Data"
    )
    .reset_index()
    .rename_axis(None, axis=1)
    .add_prefix("spend_")
    .rename(columns={"spend_State Name": "State Name"})
)
df_state_public_spend_essential_wide.head()

State Name  spend_2017  spend_2018  spend_2019  spend_2020  spend_2021
0     Alabama    56.43214       57.23       54.25       46.81       52.35
1      Alaska   114.43666       96.52       98.25      215.15       90.17
2     Arizona     8.83289        9.66       15.25       24.34       15.32
3    Arkansas    52.29329       50.57       48.23       45.90       42.94
4  California    61.69589       65.71       71.89       69.87       78.47

In [11]:
#Geographic information (Lat-Long)
geocoords=df_health_outcomes_edited[['ZIP', 'longitude', 'latitude']].drop_duplicates()
geocoords.shape, geocoords.ZIP.nunique()

((32409, 3), 32409)

##### Creating combined dataset

In [12]:
# Combined_dataset_zip contains ZIP level features.
# Combined_dataset_state contains State level features.
# Combined_dataset contains all features

# Merging dataframes for ZIP level features
combined_dataset_zip = pd.merge(
    pd.merge(
        pd.merge(
            pd.merge(
                df_income_1901_essential,
                df_health_outcomes_essential_wide,
                on="ZIP",
                how="outer",
            ),
            df_sdoh_essential_wide,
            on="ZIP",
            how="outer",
        ),
        df_acs_edited,
        on="ZIP",
        how="outer",
    ),
    df_urban_rural_edited,
    on="ZIP",
    how="left",
)

# Merging dataframes for State level features
combined_dataset_state = pd.merge(
    pd.merge(
        df_state_politics_edited,
        df_state_public_spend_essential_wide,
        on="State Name",
        how="outer",
    ),
    df_zip2st_edited,
    on="State",
    how="outer",
)

# Merging ZIP and State level features
combined_dataset = pd.merge(
    pd.merge(combined_dataset_zip, combined_dataset_state, on="ZIP", how="outer"),
    geocoords,
    on="ZIP",
    how="left",
)

# Print the shape of the combined dataset
combined_dataset.shape

(39503, 84)

In [13]:
combined_dataset.sample(5)

ZIP  Estimate Households Total  Percent Households lt 10k  \
33064  98272                     9977.0                        2.1   
26653  75681                     1073.0                       10.2   
36141  23341                        NaN                        NaN   
13558  41091                     7865.0                        1.6   
24686  70515                     1164.0                        8.1   

       Estimate Households Median income (dollars)  Estimate Families Total  \
33064                                     109051.0                   7853.0   
26653                                      58125.0                    692.0   
36141                                          NaN                      NaN   
13558                                     131442.0                   6658.0   
24686                                      47344.0                    799.0   

       Estimate Families Median income (dollars)  \
33064                                   116513.0   
26653                                    76667.0   
36141                                        NaN   
13558                                   142346.0   
24686                                    49803.0   

       Estimate Nonfamily households Total  \
33064                               2124.0   
26653                                381.0   
36141                                  NaN   
13558                               1207.0   
24686                                365.0   

       Estimate Nonfamily households Median income (dollars)  \
33064                                            69063.0       
26653                                            29453.0       
36141                                                NaN       
13558                                            51164.0       
24686                                            24970.0       

       Percent Households that are Families  TotalPopulation_x  \
33064                              0.787110            27942.0   
26653                              0.644921             2562.0   
36141                                   NaN                NaN   
13558                              0.846535            16873.0   
24686                              0.686426             3838.0   

       All Teeth Lost  Annual Checkup  Any Disability  Arthritis  \
33064            11.3            61.3            25.8       24.2   
26653            18.4            74.0            36.2       30.2   
36141             NaN             NaN             NaN        NaN   
13558             9.7            74.6            26.0       25.6   
24686            26.0            75.2            42.5       27.4   

       Binge Drinking  COPD  Cancer (except skin)  Cervical Cancer Screening  \
33064            17.6   5.1                   5.8                       82.1   
26653            15.8   9.7                   7.8                       78.4   
36141             NaN   NaN                   NaN                        NaN   
13558            17.8   5.6                   6.2                       85.3   
24686            18.9  10.7                   5.7                       76.4   

       Cholesterol Screening  Chronic Kidney Disease  Cognitive Disability  \
33064                   82.2                     2.4                  13.4   
26653                   84.2                     3.8                  15.4   
36141                    NaN                     NaN                   NaN   
13558                   87.5                     2.2                  11.6   
24686                   78.3                     3.3                  23.4   

       Colorectal Cancer Screening  Core preventive services for older men  \
33064                         67.9                                    47.9   
26653                         65.2                                    37.7   
36141                          NaN                                     NaN   
13558                         75.4                                    

In [14]:
combined_dataset['longitude'].isnull().sum()

7094

In [15]:
#Saving combined dataset
combined_dataset.to_pickle(output_path+"\\combined_dataset.pkl")
combined_dataset.to_csv(output_path+"\\combined_dataset.csv")

##### Checking if combined_dataset has values for all ZIPs

In [16]:
# % Missing values for each column. Function defined in helper functions notebook
perc_missing_num(combined_dataset)

Number of records in dataset: 39503
Missing records summary:


column  number_missing  \
0                                                 ZIP               3   
74                                              State             132   
73                                         State Name             132   
81                                               City             135   
79                                         spend_2020             319   
78                                         spend_2019             319   
77                                         spend_2018             319   
76                                         spend_2017             319   
75  Political Affiliation (2008-2020 presidential ...             319   
80                                         spend_2021            2399   
4                             Estimate Families Total            5729   
6                 Estimate Nonfamily households Total            5729   
1                           Estimate Households Total            5729   
69                      Estimate!!Total housing units            6383   
57            Estimate!!SEX AND AGE!!Total population            6383   
59                    Percent!!RACE!!Total population            6383   
8                Percent Households that are Families            6685   
2                           Percent Households lt 10k            6685   
67  Percent!!HISPANIC OR LATINO AND RACE!!Total po...            6754   
66  Percent Margin of Error!!RACE!!Total populatio...            6754   
65  Percent!!RACE!!Total population!!One race!!Nat...            6754   
64  Percent Margin of Error!!RACE!!Total populatio...            6754   
63  Percent!!RACE!!Total population!!One race!!Ame...            6754   
62  Percent Margin of Error!!RACE!!Total populatio...            6754   
61  Percent!!RACE!!Total population!!One race!!Bla...            6754   
60          Percent!!RACE!!Total population!!One race            6754   
68  Percent Margin of Error!!HISPANIC OR LATINO AN...            6760   
58  Estimate!!SEX AND AGE!!Total population!!Sex r...            6938   
43                                     Sleep <7 hours            7094   
82                                          longitude            7094   
83                                           latitude            7094   
27                                       Dental Visit            7094   
9                                   TotalPopulation_x            7094   
17                          Cervical Cancer Screening            7103   
54        Persons of racial or ethnic minority status            7167   
52                     Persons aged 65 years or older            7167   
47                                  TotalPopulation_y            7167   
21                        Colorectal Cancer Screening            7194   
51  No high school diploma among adults aged 25 ye...            7208   
36                                        Mammography            7253   
10                                     All Teeth Lost            7273   
22             Core preventive services for older men            7315   
56  Unemployment among people 16 years and older i...            7319   
23           Core preventive services for older women            7322   
53     Persons living below 150% of the poverty level            7460   
48                       Crowding among housing units            7504   
49               Housing cost burden among households            7504   
50  No broadband internet subscription among house...            7504   
55                           Single-parent households            7504   
70                                        Urban+Rural            7909   
71                                              Urban            7909   
72                                      Percent_Urban            7971   
15                                               COPD            8068   
11                                     Annual Checkup            8068   
20                              

##### Dataset with strict removal of any missing value except % Urban

In [17]:
# Creating the dataset
combined_dataset_nona_strict = combined_dataset.dropna(
    subset=combined_dataset.columns.difference(
        ["Percent_Urban", "Urban+Rural", "Urban"]
    )
)

In [18]:
# % Missing values for each column. Function defined in helper functions notebook
#perc_missing_num(combined_dataset_nona_strict)

In [19]:
combined_dataset_nona_strict.shape

(24083, 84)

##### Checking which states are completely absent in the new dataset

In [20]:
# List of states in original dataset and dataset without any missing information
states_in_original=combined_dataset['State Name'].unique()
states_in_nona_strict=combined_dataset_nona_strict['State Name'].unique()

In [21]:
print("States not present in nona but present in original dataset")
for state in states_in_original:
    if state not in states_in_nona_strict: print(state)

States not present in nona but present in original dataset
Puerto Rico
nan
Rhode Island
Delaware
West Virginia
Florida
Kansas
Utah
American Samoa
Northern Mariana Islands
Trust Territories
Guam
Virgin Islands


In [22]:
#combined_dataset[combined_dataset['State Name'].isin(['Kansas', 'Utah', 'Florida', 'Delaware', 'West Virginia'])].sample(20)
print("# unique values for spend 2021 for Utah, Kansas, Delaware, and West Virginia:", 
      combined_dataset[combined_dataset['State Name'].isin(['Utah', 'Kansas', 'Delaware', 'West Virginia'])]['spend_2021'].unique())
print("# unique values for Annual Checkup for Florida:", 
      combined_dataset[combined_dataset['State Name'].isin(['Florida'])]['Annual Checkup'].unique())

# Note: Kansas, Utah, Delaware, and West Virginia does not have Spend 2021 variable populated. 
## Note2: Florida doesn't have majority of health outcomes populated.


# unique values for spend 2021 for Utah, Kansas, Delaware, and West Virginia: [nan]
# unique values for Annual Checkup for Florida: [nan]


##### Dataset with removal of any missing value after removing Spend 2021 feature

In [23]:
# It seems resonable to drop 'spend_2021' to get Kansas, Utah, Delaware, and West Virginia back in the data.
# However, I can't see a path to include Florida.
# We can probably take it as a separate case and verify our findings from rest of the data.
# Giving -99 values to missing urban_percentage as we don't want to drop ZIPs just because of this column
combined_dataset_nona = combined_dataset.drop(columns=["spend_2021"]).dropna(
    subset=combined_dataset.columns.difference(
        ["Percent_Urban", "Urban+Rural", "Urban", "spend_2021"]
    )
)
combined_dataset_nona.shape

(25315, 83)

In [24]:
# % Missing values for each column. Function defined in helper functions notebook
#perc_missing_num(combined_dataset_nona)

In [25]:
states_in_nona=combined_dataset_nona['State Name'].unique()
print("States not present in nona but present in original dataset")
for state in states_in_original:
    if state not in states_in_nona: print(state)

States not present in nona but present in original dataset
Puerto Rico
nan
Florida
American Samoa
Northern Mariana Islands
Trust Territories
Guam
Virgin Islands


In [26]:
#Summary of #ZIPs by State to see national coverage. Unfortunately Florida is not included
#combined_dataset_nona[['State Name', 'ZIP']].groupby(by='State Name').count()

In [27]:
# Creating a separate dataframe for Florida with only those features that are available for Florida
# Giving -99 values to missing urban_percentage as we don't want to drop ZIPs just because of this column
fl_allna_dropped = combined_dataset[
    combined_dataset["State Name"].isin(["Florida"])
].dropna(axis=1, how="all")
combined_dataset_nona_florida = fl_allna_dropped.dropna(
    subset=fl_allna_dropped.columns.difference(
        ["Percent_Urban", "Urban+Rural", "Urban", "spend_2021"]
    )
)
combined_dataset_nona_florida.shape

(892, 55)

Insights:
<ol>
<li>29 health outcomes features don't have data for 8,068 or 20% ZIPs. Most of these ZIPs should be 0 or low population as per Census procedure documented in the notes for this dataset.</li>
<li>Another 21 features don't have data for at least 7,000 or more ZIPs.</li>
<li>Another 18 features don't have data for at least 5,704 or more ZIPs.</li>
<li>Removing all ZIPs with missing data leaves us with 24,089 ZIPs (61%) out of 39,503 ZIPs in the original combined dataset. But it will exclude Kansas, Utah and Florida all together.</li>
<li>Kansas, Utah, Delaware and West Virginia do not have Spend 2021 variable populated. Florida doesn't have majority of health outcomes populated.</li>
<li>Recommend to  drop spend 2021 to get Kansas, Utah, Delaware and West Virginia back in the data. </li>
<li>However, I can't see a path to include Florida. We can probably take it as a separate case and verify our findings from rest of the data.</li>
<li>This leaves us with 25,321 ZIPs with 80 features in the primary dataset and 893 ZIPs with 51 features in the Florida dataset. </li>
<li>This seems like a reasonable sample size with national coverage. Dropping ZIPs with missing data will remove the noise.</li>
<li>Recommend to move ahead with 25K ZIPs for further analysis.</li>
</ol>

##### Saving dataset without any missing information for any ZIP

In [28]:
#Saving dataset without any missing information
combined_dataset_nona.to_pickle(output_path+"\\combined_dataset_nona.pkl")
combined_dataset_nona.to_csv(output_path+"\\combined_dataset_nona.csv")

#Saving separate dataset for Florida
combined_dataset_nona_florida.to_pickle(output_path+"\\combined_dataset_nona_florida.pkl")
combined_dataset_nona_florida.to_csv(output_path+"\\combined_dataset_nona_florida.csv")

In [29]:
#25321 ZIPs / 82 cols in rest of US, 893 ZIPs / 54 cols in florida
combined_dataset_nona.shape, combined_dataset_nona_florida.shape

((25315, 83), (892, 55))

In [30]:
%watermark -u -i -d -m -v -iv -w

Last updated: 2024-02-14T23:13:16.544830-05:00

Python implementation: CPython
Python version       : 3.11.7
IPython version      : 8.20.0

Compiler    : MSC v.1916 64 bit (AMD64)
OS          : Windows
Release     : 10
Machine     : AMD64
Processor   : Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
CPU cores   : 8
Architecture: 64bit

numpy     : 1.26.3
plotly    : 5.9.0
re        : 2.2.1
matplotlib: 3.8.0
pandas    : 2.1.4
seaborn   : 0.13.2

Watermark: 2.4.3

